File: dsas-code/LeetCode/Partition Labels/Partition Labels.py

class Solution:
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        p = []
        for k in range(len(S)):
            # p中无含有S[k]的字符串
            flg = False
            index = -1
            for i in range(len(p)):
                if S[k] in p[i]:
                    # p中有含有S[k]的字符串
                    index = i
                    flg = True
                    break
            p.append(S[k])
            if flg:
                # 合并index到len(p)的字符串
                for i in range(index+1, len(p)):
                    p[index] += p[i]
                p = p[0:index+1]
        num = []
        for k in p:
            num.append(len(k))
        return num


if __name__ == "__main__":
    su = Solution()
    print(su.partitionLabels("ababcbacadefegdehijhklij"))

File: dsas-code/LeetCode/Partition Labels/Partition Labels.cpp

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> partitionLabels(string S) {
        std::vector<int> vi;
        std::vector<std::string> vs;
        for (char c:S)
        {
            bool flg = false;
            unsigned int i;
            for (i = 0; i < vs.size(); i ++)
            {
                if (std::string::npos != vs[i].find(c))
                {
                    flg = true;
                    break;
                }
            }
            vs.push_back(std::string(""));
            vs[vs.size()-1] += c;
            if (flg)
            {
                int size = vs.size();
                for (int j = i + 1; j < size; j ++)
                    vs[i] += vs[j];
                for (int j = i + 1; j < size; j ++)
                    vs.pop_back();
            }
        }
        for (auto it:vs)
            vi.push_back(it.size());
        return vi;
    }
};

int main(void)
{
    Solution su;
    std::vector<int> vi = su.partitionLabels("ababcbacadefegdehijhklij");
    for (auto k:vi)
        std::cout << k << std::endl;
    return 0;
}
